cardano-api/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,395 @@
+    xMetadata === yMetadata
+
+
+-- * Normalization
+--
+-- Strip unnecessary details
+--
+-- In many cases, after roundtrip,
+-- @Just mempty@ may become @Nothing@ or vice versa.
+-- Input data also may be generated as either @Just 0@ or @Nothing@.
+-- Order of some items may also change, they need to be reordered.
+
+-- | Normalizations applied to original body content
+normalizeContentOriginal :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeContentOriginal content =
+  content
+    { txAuxScripts = normalizeAuxScripts $ txAuxScripts content
+    , txCertificates = normalizeCertificates $ txCertificates content
+    , txIns = sortOn fst $ txIns content
+    , txInsCollateral = normalizeInsCollateral $ txInsCollateral content
+    , txMetadata = normalizeMetadata $ txMetadata content
+    , txMintValue = normalizeMintValue $ txMintValue content
+    , txWithdrawals = normalizeWithdrawals $ txWithdrawals content
+    }
+
+-- | Normalizations applied to roundtrip result body content
+normalizeContentRoundtrip ::
+  TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeContentRoundtrip content =
+  content
+    { txAuxScripts = normalizeAuxScripts $ txAuxScripts content
+    , txIns = sortOn fst $ txIns content
+    , txInsCollateral = normalizeInsCollateral $ txInsCollateral content
+    }
+
+-- | Original data: Unify empty and None.
+--
+-- Roundtrip data: Sort transaction input ids.
+normalizeInsCollateral :: TxInsCollateral era -> TxInsCollateral era
+normalizeInsCollateral = \case
+  TxInsCollateralNone -> TxInsCollateralNone
+  TxInsCollateral support ins
+    -- for original:
+    | null ins -> TxInsCollateralNone
+    -- for roundtrip:
+    | otherwise -> TxInsCollateral support $ sort ins
+
+-- | Unify empty and None.
+normalizeMetadata :: TxMetadataInEra era -> TxMetadataInEra era
+normalizeMetadata = \case
+  TxMetadataInEra _ (TxMetadata m) | null m -> TxMetadataNone
+  other -> other
+
+-- | Unify empty and None.
+-- Upgrade script versions (see Upgrading scripts section).
+-- Stabilize order of scripts sorting them by language.
+normalizeAuxScripts :: TxAuxScripts era -> TxAuxScripts era
+normalizeAuxScripts = \case
+  TxAuxScripts _ [] -> TxAuxScriptsNone
+  TxAuxScripts support scripts ->
+    -- sorting uses script versions, hence sort after upgrade
+    TxAuxScripts support $
+    sortBy normalizeScriptInEraOrder $
+    map upgradeScriptInEra scripts
+  other -> other
+
+normalizeScriptInEraOrder :: ScriptInEra era -> ScriptInEra era -> Ordering
+normalizeScriptInEraOrder
+  (ScriptInEra slang1 script1)
+  (ScriptInEra slang2 script2) =
+    case testEquality lang1 lang2 of
+      Nothing -> compare (AnyScriptLanguage lang1) (AnyScriptLanguage lang2)
+      Just Refl -> normalizeScriptOrder script1 script2
+    where
+      lang1 = languageOfScriptLanguageInEra slang1
+      lang2 = languageOfScriptLanguageInEra slang2
+
+normalizeScriptOrder :: Script lang -> Script lang -> Ordering
```

Comment:
Why not implement an `Ordering` instance for `Script lang`? This is hacky.

Reply:
I don't think this instance is natural enough to expose it in the API. It is used here only to compare script sets, not to order them.

Reply:
So this is why it is named "normalize order", not "sort".

I'll add it to its doc.